Introduce SelectiveStreamReader

mbasmanova opened the pull request:
Depends on #12938

Part of #12585

highker commented on presto-orc/src/test/java/com/facebook/presto/orc/OrcTester.java:
```diff
@@ -510,51 +523,52 @@ private void assertRoundTrip(Type writeType, Type readType, List<?> writeValues,
                 rowsProcessed += batchSize;
             }
-            assertFalse(iterator.hasNext());
+            assertEquals(rowsProcessed, expectedValues.get(0).size());
```
maybe
```java
int actualRowsProcessed = rowsProcessed;
expectedValues.forEach(value -> assertEquals(actualRowsProcessed, value.size()));

mbasmanova replied:
I'll add a sanity check at the beginning of the method:

```
        int expectedRowsProcessed = expectedValues.get(0).size();
        expectedValues.forEach(values -> assertEquals(values.size(), expectedRowsProcessed));
```

and change this to

```
assertEquals(rowsProcessed, expectedRowsProcessed);
```

highker commented on presto-orc/src/test/java/com/facebook/presto/orc/OrcTester.java:
```diff
@@ -826,19 +860,31 @@ private static void assertFileContentsDwrfHive(
+        assertEquals(rowCount, expectedValues.get(0).size());
+    }
+
+    private static List<String> makeColumnNames(List<Type> types)
```
nit: `private static List<String> makeColumnNames(int columns)`

highker commented on presto-orc/src/test/java/com/facebook/presto/orc/OrcTester.java:
```diff
@@ -956,19 +1002,61 @@ private static Object decodeRecordReaderMap(Type type, Map<?, ?> map)
+
     public static DataSize writeOrcFileColumnHive(File outputFile, Format format, RecordWriter recordWriter, Type type, Iterator<?> values)
             throws Exception
     {
```
I guess the body of this function could be something like
```java
ImmutableList.Builder builder = ImmutableList.builder();
values.forEachRemaining(builder::add);
return writeOrcFileColumnHive(outputFile, format, recordWriter, ImmutableList.of(type), ImmutableList.of(builder.build()));

highker commented on presto-orc/src/test/java/com/facebook/presto/orc/OrcTester.java:
```diff
@@ -956,19 +1002,61 @@ private static Object decodeRecordReaderMap(Type type, Map<?, ?> map)
     public static DataSize writeOrcColumnHive(File outputFile, Format format, CompressionKind compression, Type type, Iterator<?> values)
             throws Exception
     {
         RecordWriter recordWriter;
```
body could be (or something similiar)
```java
ImmutableList.Builder builder = ImmutableList.builder();
values.forEachRemaining(builder::add);
return writeOrcColumnsHive(outputFile, format, compression, ImmutableList.of(type), ImmutableList.of(builder.build()));

highker commented on presto-orc/src/test/java/com/facebook/presto/orc/OrcTester.java:
```diff
@@ -510,51 +523,52 @@ private void assertRoundTrip(Type writeType, Type readType, List<?> writeValues,
             throws IOException
     {
-        try (OrcRecordReader recordReader = createCustomOrcRecordReader(tempFile, orcEncoding, createOrcPredicate(type, expectedValues, format, isHiveWriter), type, MAX_BATCH_SIZE)) {
+        OrcPredicate orcPredicate = types.size() > 1 ? TRUE : createOrcPredicate(types.get(0), expectedValues.get(0), format, isHiveWriter);
```
I think instead of skipping the rest of the predicates, there is a quite fix to `TestingOrcPredicate` to enforce predicates for all columns. All predicate classes in `TestingOrcPredicate` inherit `BasicOrcPredicate`. Instead of having `List<T> expectedValues` in `BasicOrcPredicate`, we could make it a list of list like what you have changed for `OrcTester` class. The input `statisticsByColumnIndex` should have the exact same size as `expectedValues`. We can assert column stats one after another.

mbasmanova replied:
@highker You are right. Looks like it is quite straightforward to add support for multiple columns to TestingOrcPredicate. I put the changes into a fixup commit hoping it would be easier to review: https://github.com/prestodb/presto/pull/12991/commits/15c5f09328db769aae58937d25fcaa244195dade

mbasmanova commented:
@highker James, thank you for review. Comments addressed.

mbasmanova commented:
@highker 

> "fixup! Extend OrcTester to support multi-column files" LGTM. It would be nice to squash it with "Extend OrcTester to support multi-column files". 

Thanks. Squashed.

highker commented on presto-orc/src/main/java/com/facebook/presto/orc/reader/BaseStreamReader.java:
```diff
@@ -0,0 +1,33 @@
+import java.io.IOException;
+import java.util.List;
+
+public interface BaseStreamReader
```
Having both `StreamReader` and `SelectiveStreamReader` inheriting `BaseStreamReader` is a bit confusing from the naming. Can we actually have:
- `BaseStreamReader ` -> `StreamReader`
- `StreamReader` -> `BatchStreamReader`
- `SelectiveStreamReader` -> `SelectiveStreamReader`

It would be easier to understand the behavior of the two stream readers. Also making all inheriting classes of `BatchStreamReader` having `Batch` in the name would be helpful to understand.

highker commented on presto-orc/src/main/java/com/facebook/presto/orc/OrcRecordReader.java:
```diff
@@ -138,229 +75,39 @@ public OrcRecordReader(
-        return presentColumns.contains(hiveColumnIndex);
     }
 
     public int nextBatch()
```
I think some of the function body can be extracted out as a common part for both readers. That can prevent many variables becoming "protected":
```java
int prepareNextBatch()
        throws IOException
{
    // update position for current row group (advancing resets them)
    filePosition += currentBatchSize;
    currentPosition += currentBatchSize;

    // if next row is within the current group return
    if (nextRowInGroup >= currentGroupRowCount) {
        // attempt to advance to next row group
        if (!advanceToNextRowGroup()) {
            filePosition = fileRowCount;
            currentPosition = totalRowCount;
            return -1;
        }
    }

    // We will grow currentBatchSize by BATCH_SIZE_GROWTH_FACTOR starting from initialBatchSize to maxBatchSize or
    // the number of rows left in this rowgroup, whichever is smaller. maxBatchSize is adjusted according to the
    // block size for every batch and never exceed MAX_BATCH_SIZE. But when the number of rows in the last batch in
    // the current rowgroup is smaller than min(prepareNextBatch, maxBatchSize), the prepareNextBatch for next batch in
    // the new rowgroup should be grown based on min(prepareNextBatch, maxBatchSize) but not by the number of rows in
    // the last batch, i.e. currentGroupRowCount - nextRowInGroup. For example, if the number of rows read for
    // single fixed width column are: 1, 16, 256, 1024, 1024,..., 1024, 256 and the 256 was because there is only
    // 256 rows left in this row group, then the prepareNextBatch should be 1024 instead of 512. So we need to grow the
    // prepareNextBatch before limiting the currentBatchSize by currentGroupRowCount - nextRowInGroup.
    currentBatchSize = toIntExact(min(nextBatchSize, maxBatchSize));
    nextBatchSize = min(currentBatchSize * BATCH_SIZE_GROWTH_FACTOR, MAX_BATCH_SIZE);
    currentBatchSize = toIntExact(min(currentBatchSize, currentGroupRowCount - nextRowInGroup));

    return currentBatchSize;
}

highker replied:
Also, this `prepareNextBatch` can return a struct containing both `currentBatchSize` and `nextRowInGroup` (with `nextRowInGroup` updated) so that we can avoid `nextRowInGroup` becoming "protected"

highker commented on presto-orc/src/main/java/com/facebook/presto/orc/BaseOrcRecordReader.java:
```diff
@@ -0,0 +1,610 @@
+    private final AggregatedMemoryContext systemMemoryUsage;
+
+    private final Optional<OrcWriteValidation> writeValidation;
+    protected final Optional<OrcWriteValidation.WriteChecksumBuilder> writeChecksumBuilder;
```
These validation flags are not needed for selective readers. The purpose of these flags are to verify if the orc writer writes the correct data. It does not matter what reader it uses. The validation part does not have to be in the selective reader. We could move these flags to the old reader.

mbasmanova replied:
@highker Eventually, once the new readers are fully-functional and robust, I'd like to delete the current readers.

highker commented on presto-orc/src/main/java/com/facebook/presto/orc/OrcRecordReader.java:
```diff
@@ -419,111 +166,14 @@ public Block readBlock(Type type, int columnIndex)
         return block;
```
The function `readBlock` contains pretty critical logic in terms of deciding the next batch size. Do we wanna do something similar to the selective reader? In general, I'm curious how selective reader can control memory consumption given the load size is quite unpredictable? Both Facebook and Twitter see severe reliability issues (e.g., humongous allocation) due to unpredictable load size. cc: @wenleix

mbasmanova replied:
@highker Indeed, batch size adaptation is going to be non-trivial, hence, it is not included in this PR. At a high level, we'll be monitoring filter selectivity and average size of records that pass the filters and adjust the batch size accordingly.

highker commented on presto-orc/src/main/java/com/facebook/presto/orc/BaseOrcRecordReader.java:
```diff
@@ -0,0 +1,610 @@
+import static java.util.Comparator.comparingLong;
+import static java.util.Objects.requireNonNull;
+
+abstract class BaseOrcRecordReader
```
Same, maybe:
- `BaseOrcRecordReader` -> `AbstractOrcRecordReader`
- `OrcRecordReader` -> `OrcBatchRecordReader`
- `OrcSelectiveRecordReader` -> `OrcSelectiveRecordReader`

highker commented on presto-orc/src/main/java/com/facebook/presto/orc/BaseOrcRecordReader.java:
```diff
@@ -0,0 +1,610 @@
+
+    private final OrcDataSource orcDataSource;
+
+    protected final BaseStreamReader[] baseStreamReaders;
```
Interleaved protected and private variables kinda make the class a bit less abstracted. I put some comments to address the issues but not all. Hope we could keep all variables private. @wenleix, @tdcmeehan, @yingsu00 could possibly provide more ideas.

mbasmanova replied:
@highker James, thanks for the suggestion about `prepareNextBatch()`. I applied it along with a handful of other tricks and removed all `protected` member variables from `AbstractOrcRecorddReader`.

mbasmanova commented:
@highker James, thank you for reviewing and suggesting better names for the classes. I applied your comments and updated the PR.

tdcmeehan commented on presto-orc/src/main/java/com/facebook/presto/orc/OrcRecordReader.java:
```diff
@@ -138,186 +75,42 @@ public OrcRecordReader(
-        }
-        return new CachingOrcDataSource(dataSource, createTinyStripesRangeFinder(stripes, maxMergeDistance, tinyStripeThreshold));
-    }
+        batchStreamReaders = Arrays.stream(getStreamReaders())
```
Could we parameterize `AbstractOrcRecordReader`?  Like: `abstract class AbstractOrcRecordReader<T extends StreamReader>`.  That way we don't need to cast.

tdcmeehan commented on presto-orc/src/main/java/com/facebook/presto/orc/reader/StreamReader.java:
```diff
@@ -0,0 +1,33 @@
+import java.io.IOException;
+import java.util.List;
+
+public interface StreamReader
```
I can see how this interface is useful for the abstractions in this PR, but it's not clear from the methods on the interface how this is used for reading streams.  Basically I wonder if there's a better name for this interface.

mbasmanova replied:
@tdcmeehan @highker Do you have a suggestion?

tdcmeehan commented on presto-orc/src/test/java/com/facebook/presto/orc/BenchmarkStreamReaders.java:
```diff
@@ -92,7 +92,7 @@
             throws Throwable
     {
-        OrcRecordReader recordReader = data.createRecordReader();
+        OrcBatchRecordReader recordReader = data.createRecordReader();
```
Should this be renamed to `BenchmarkBatchStreamReaders`?

tdcmeehan commented on presto-orc/src/main/java/com/facebook/presto/orc/OrcReader.java:
```diff
@@ -222,12 +222,12 @@ public CompressionKind getCompressionKind()
     }
 
-    public OrcRecordReader createRecordReader(
+    public OrcBatchRecordReader createRecordReader(
```
Should this be called `createBatchRecordReader`?

mbasmanova replied:
@tdcmeehan @highker I'm fine either way, but I feel that if we rename this method, then, perhaps, we should rename all of the following as well:

- OrcPageSource -> OrcBatchPageSource
- OrcPageSourceFactory -> OrcBatchPageSourceFactory
- DwrfPageSourceFactory -> DwrfBatchPageSourceFactory
- OrcReader -> OrcBatchReader

What do you think?

highker replied:
If we end up with two different page sources, then probably yes.

mbasmanova replied:
@highker Sounds good. Would you take a look at `Introduce SelectiveStreamReader` commit. I'll do these renames while addressing feedback on that.

highker replied:
Yes, actually let me do a review before changing the names

tdcmeehan commented:
It would be nice to have a JMH benchmark that ensures the selective readers are performant.

tdcmeehan commented:
@mbasmanova there are checkstyle failures: 

```
[ERROR] src/main/java/com/facebook/presto/raptor/storage/OrcPageFileRewriter.java:[18] (imports) ImportOrder: Wrong order for 'com.facebook.presto.orc.OrcBatchRecordReader' import.
[ERROR] src/main/java/com/facebook/presto/raptor/storage/ShardStats.java:[18] (imports) ImportOrder: Wrong order for 'com.facebook.presto.orc.OrcBatchRecordReader' import.
[ERROR] src/main/java/com/facebook/presto/raptor/storage/OrcStorageManager.java:[21] (imports) ImportOrder: Wrong order for 'com.facebook.presto.orc.OrcBatchRecordReader' import.
[ERROR] src/main/java/com/facebook/presto/raptor/storage/OrcPageSource.java:[18] (imports) ImportOrder: Wrong order for 'com.facebook.presto.orc.OrcBatchRecordReader' import.
[ERROR] src/test/java/com/facebook/presto/raptor/storage/TestOrcStorageManager.java:[17] (imports) ImportOrder: Wrong order for 'com.facebook.presto.orc.OrcBatchRecordReader' import.
[ERROR] src/test/java/com/facebook/presto/raptor/storage/TestShardWriter.java:[20] (imports) ImportOrder: Wrong order for 'com.facebook.presto.orc.OrcBatchRecordReader' import.
[ERROR] src/test/java/com/facebook/presto/raptor/storage/OrcTestingUtil.java:[22] (imports) ImportOrder: Wrong order for 'com.facebook.presto.orc.OrcBatchRecordReader' import.
[ERROR] src/test/java/com/facebook/presto/raptor/storage/TestOrcFileRewriter.java:[20] (imports) ImportOrder: Wrong order for 'com.facebook.presto.orc.OrcBatchRecordReader' import.
```

mbasmanova commented:
@tdcmeehan @oerling @highker

> It would be nice to have a JMH benchmark that ensures the selective readers are performant.

Tim, thanks for the request. I added a benchmark, compared results with `BenchmarkBatchStreamReaders` and saw a regression: 1.5x for the case when there are not nulls, 8% for the case with nulls, 1.6x for the case of all nulls. To fix that I split the code into (1) all nulls; (2) no filter; (3) filter. The numbers for same use cases (e.g. no filters) now match:

```
BenchmarkSelectiveStreamReaders.readAllNull                            boolean  avgt   20  0.003 ±  0.001   s/op
BenchmarkSelectiveStreamReaders.readBooleanNoNull                          N/A  avgt   20  0.029 ±  0.001   s/op
BenchmarkSelectiveStreamReaders.readBooleanNoNullWithFilter                N/A  avgt   20  0.098 ±  0.006   s/op
BenchmarkSelectiveStreamReaders.readBooleanWithNull                        N/A  avgt   20  0.098 ±  0.004   s/op
BenchmarkSelectiveStreamReaders.readBooleanWithNullWithFilter              N/A  avgt   20  0.134 ±  0.006   s/op

BenchmarkBatchStreamReaders.readAllNull                  boolean  avgt   20  0.003 ±  0.001   s/op
BenchmarkBatchStreamReaders.readBooleanNoNull                N/A  avgt   20  0.030 ±  0.001   s/op
BenchmarkBatchStreamReaders.readBooleanWithNull              N/A  avgt   20  0.106 ±  0.002   s/op
```

mbasmanova commented on presto-orc/src/main/java/com/facebook/presto/orc/reader/SelectiveStreamReader.java:
```diff
@@ -0,0 +1,33 @@
+        extends StreamReader
+{
+    // Extract values at the specified positions, apply filter and buffer the values that pass the filter.
+    // Return the number of positions that passed the filter.
```
@yingsu00 Fixed and beefed up the documentation on this interface.

mbasmanova commented on presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java:
```diff
@@ -0,0 +1,192 @@
+            List<Integer> outputColumns,
+            Map<Integer, TupleDomainFilter> filters,
+            Map<Integer, List<Subfield>> requiredSubfields,
+            Map<Integer, Object> constantValues,
```
Not yet. It is needed to apply filters on partition key, $bucket and $path columns as well as columns that are missing from the file as a result of schema evolution.

mbasmanova commented on presto-hive/src/main/java/com/facebook/presto/hive/orc/OrcSelectivePageSourceFactory.java:
```diff
@@ -0,0 +1,352 @@
+        return new TupleDomainOrcPredicate<>(compactEffectivePredicate, columnReferences.build(), orcBloomFiltersEnabled);
+    }
+
+    private static TupleDomain<HiveColumnHandle> toCompactTupleDomain(TupleDomain<HiveColumnHandle> effectivePredicate, int threshold)
```
@highker I moved this logic into `TupleDomainOrcPredicate`

mbasmanova commented on presto-orc/src/main/java/com/facebook/presto/orc/reader/BooleanSelectiveStreamReader.java:
```diff
@@ -0,0 +1,332 @@
+
+        boolean nullsAllowed = filter == null || filter.testNull();
+        boolean recordNulls = nullsAllowed && presentStream != null;
+        if (outputRequired) {
```
That's right. These arrays are not needed for filtering; only for producing output blocks.

mbasmanova commented:
@highker @tdcmeehan @yingsu00  James, Tim, Ying, thank you for the feedback. All addressed and ready for another round.

highker commented on presto-orc/src/main/java/com/facebook/presto/orc/reader/BooleanSelectiveStreamReader.java:
```diff
@@ -0,0 +1,332 @@
+                    streamPosition = position;
+                }
+
+                if (presentStream != null && !presentStream.nextBit()) {
```
The reason I was curious about https://github.com/prestodb/presto/pull/12991/commits/77178e4947d85560cecdc0e9ba092f4e1f1c2c3f#r298004032 is because there are a few if/elses in this logic. Some of them need some thinking before understanding what it does exactly. Possibly it's due to the different combination of `null X filter X outputRequired`. Can we extract some conditions and assignments into helper functions? For example, `if (presentStream != null && !presentStream.nextBit())` could be `private boolean nextPositionIsNull()`. Some of the if/else checks could be hidden in the helper functions for easy understanding.

mbasmanova replied:
- `filter == null` implies `outputRequired == true`, because if there is no filter and no output is needed, then the reader should not have been created in the first place
- `nextPositionIsNull()` will have a side effect of advancing the presentStream; are we OK with that?
- I agree that the are a bunch of flags and combinations that need to be handled carefully

mbasmanova replied:
Added `checkArgument(filter.isPresent() || outputRequired, "filter must be present if outputRequired is false");` to the constructor of the `BooleanSelectiveStreamReader`.

highker commented on presto-orc/src/main/java/com/facebook/presto/orc/reader/BooleanSelectiveStreamReader.java:
```diff
@@ -0,0 +1,332 @@
+                }
+
+                if (presentStream != null && !presentStream.nextBit()) {
+                    nulls[i] = true;
```
Do we need to check `outputRequired` and `recordNulls` before doing the assignment?

mbasmanova replied:
- here we have `filter == null`, hence `outputRequired == true`
- we also have a null value which means we have to record; sometimes we don't need to record non-null values, but we always record nulls (except when all values are null, but that case is processed separately)
- would it help to add some comments throughout this code?

highker commented on presto-orc/src/main/java/com/facebook/presto/orc/reader/BooleanSelectiveStreamReader.java:
```diff
@@ -0,0 +1,332 @@
+            else {
+                outputPositionCount = 0;
+            }
+        }
```
We might need to check
```java
else if (dataStream == null) {
    throw new OrcCorruptionException(streamDescriptor.getOrcDataSourceId(), "Value is not null but data stream is not present");
}

mbasmanova replied:
dataStream is null in this case... did you think of some other branch?

highker commented on presto-orc/src/main/java/com/facebook/presto/orc/reader/BooleanSelectiveStreamReader.java:
```diff
@@ -0,0 +1,332 @@
+
+                if (presentStream != null && !presentStream.nextBit()) {
+                    if (nullsAllowed) {
+                        if (outputRequired) {
```
Do we wanna check `recordNulls`? There are too many flag combinations. Having helpers would be easier to understand.

mbasmanova replied:
- `isNull == true` must be recorded all the time; only `isNull == false` recording may be unnecessary
- would it help to rename `recordNulls` to `recordNotNull`?

mbasmanova commented:
@tdcmeehan @highker While renaming `BenchmarkStreamReaders` I noticed BenchmarkOrcDecimalReader which is the same as `BenchmarkStreamReaders`, but for long decimal type. I ended up folding that benchmark into `BenchmarkStreamReaders`. Hence, a few extra commits:

```
b923222b38 Fold BenchmarkOrcDecimalReader into BenchmarkBatchStreamReaders
4910e701be Rename *Decimal* to *ShortDecimal* in BenchmarkBatchStreamReaders
b9dbde625d Refactor BenchmarkBatchStreamReaders
```

highker commented:
@mbasmanova benchmark looks good to me. Thanks for the heads up.

mbasmanova commented on presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java:
```diff
@@ -0,0 +1,192 @@
+        batchRead(batchSize);
+
+        if (positionCount == 0) {
+            return new Page(0);
```
`null` means there is no MORE data; here `positionCount == 0` means that in this batch there were no records that passed all filters

highker commented:
@wenleix, because it reads data in batches.

wenleix commented:
BTW the commit organization of this PR looks great! 😃

wenleix commented:
> @wenleix, because it reads data in batches.

@highker  But doesn't selective reader also read in batch? e.g. https://github.com/prestodb/presto/pull/12991/commits/9291dc9a27bc02b0e243f55f04caa179dc4dd56e#diff-4a9f3483e806b7421dcd0c57144598f9R142.

highker commented:
@wenleix, the old reader will load all rows once the batch size is fixed. The new reader only estimate a batch size but selectively load data.

tdcmeehan commented on presto-orc/src/main/java/com/facebook/presto/orc/reader/BooleanSelectiveStreamReader.java:
```diff
@@ -0,0 +1,332 @@
+        }
+
+        byte[] values = new byte[positionCount];
+        boolean[] nulls = null;
```
It's confusing to give this the same name as the member variable

mbasmanova replied:
How about `valuesCopy` and `nullsCopy`?

tdcmeehan replied:
OK

mbasmanova replied:
@tdcmeehan Renamed to `xxxCopy`.

mbasmanova commented:
@highker Thank you for reviewing, James!

tdcmeehan commented on presto-orc/src/main/java/com/facebook/presto/orc/reader/BooleanSelectiveStreamReader.java:
```diff
@@ -0,0 +1,332 @@
+                continue;
+            }
+
+            assert outputPositions[i] == nextPosition;
```
Nice!

yingsu00 replied:
So we're favoring assert to verify in the future? @mbasmanova @tdcmeehan

tdcmeehan replied:
I think it makes sense in this particular case as it's a hot path loop and we'd like to make the verification optional during production runs.

yingsu00 commented:
> @wenleix, because it reads data in batches.

But the SelectiveStreamReader also read in batches.

tdcmeehan commented:
@mbasmanova I found the `BooleanSelectiveStreamReader` complicated and confusing to read.  Given that in the future there will be more of these readers, I tried to see if I could simplify it.  What I did was tread the `read` method like a static machine, where it just checks the current state then delegates to different combinations of states: no data (all nulls), data present but no filter, data present with filter and nulls, data present with filter but without nulls.  The result is it's around 50 LOCs bigger, but in my opinion this is easier to follow, and I don't think we sacrifice performance.  Here's a gist of my refactor: https://gist.github.com/tdcmeehan/166cab9cda977cd1ef2cc973a8cd3eab

Thoughts?

mbasmanova commented on presto-hive/src/main/java/com/facebook/presto/hive/HivePageSourceProvider.java:
```diff
@@ -120,6 +129,63 @@ public ConnectorPageSource createPageSource(ConnectorTransactionHandle transacti
+            }
+        }
+
+        throw new RuntimeException("Could not find a file reader for split " + split);
```
Sure. I'll change the other (pre-existing) instance as well.

mbasmanova commented on presto-hive/src/main/java/com/facebook/presto/hive/orc/OrcSelectivePageSourceFactory.java:
```diff
@@ -0,0 +1,312 @@
+
+            OrcSelectiveRecordReader recordReader = reader.createSelectiveRecordReader(
+                    columnTypes,
+                    Lists.transform(outputColumns, indexMapping::get),
```
Will change to `outputColumns.stream().map(indexMapping::get).collect(toImmutableList())`

tdcmeehan replied:
Sounds good

mbasmanova commented on presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java:
```diff
@@ -0,0 +1,97 @@
+        assertQueryUsingH2Cte("SELECT COUNT(*) FROM lineitem_ex WHERE ship_by_air is not null AND is_returned = true");
+    }
+
+    private void assertQueryUsingH2Cte(String query)
```
I did this because the CTE is added only to H2 query and the CTE itself is H2 specific. Would you prefer to rename to `assertQueryUsingCte`?

tdcmeehan replied:
Sounds good

mbasmanova commented:
@tdcmeehan 

> What I did was tread the read method like a state machine, where it just checks the current state then delegates to different combinations of processing the input depending on the state: no data (all nulls), data present but no filter, data present with filter and nulls, data present with filter but without nulls. 

This looks nice. I tried this approach, but BenchmarkSelectiveStreamReaders.readBooleanNoNull regressed from 0.029 s/op to 0.050 s/op. I also noticed that if I refactor only 2 branches of code (all nulls and no filter), the benchmark doesn't show any regression.

Here is what I have now. Do you like this better? (I'm concerned that this shape would prompt readers to refactor the last branch and if they don't run the benchmark they won't know that it introduced a regression.)

```
        int streamPosition = 0;
        if (dataStream == null && presentStream != null) {
            streamPosition = readAllNulls(positions, positionCount);
        }
        else if (filter == null) {
            streamPosition = readNoFilter(positions, positionCount);
        }
        else {
            outputPositionCount = 0;
            for (int i = 0; i < positionCount; i++) {
                int position = positions[i];
                if (position > streamPosition) {
                    skip(position - streamPosition);
                    streamPosition = position;
                }

                if (presentStream != null && !presentStream.nextBit()) {
                    if (nullsAllowed) {
                        if (outputRequired) {
                            nulls[outputPositionCount] = true;
                        }
                        outputPositions[outputPositionCount] = position;
                        outputPositionCount++;
                    }
                }
                else {
                    boolean value = dataStream.nextBit();
                    if (filter.testBoolean(value)) {
                        if (outputRequired) {
                            values[outputPositionCount] = (byte) (value ? 1 : 0);
                            if (presentStream != null && nullsAllowed) {
                                nulls[outputPositionCount] = false;
                            }
                        }
                        outputPositions[outputPositionCount] = position;
                        outputPositionCount++;
                    }
                }
                streamPosition++;
            }
        }
```

tdcmeehan commented:
> @tdcmeehan
> 
> > What I did was tread the read method like a state machine, where it just checks the current state then delegates to different combinations of processing the input depending on the state: no data (all nulls), data present but no filter, data present with filter and nulls, data present with filter but without nulls.
> 
> This looks nice. I tried this approach, but BenchmarkSelectiveStreamReaders.readBooleanNoNull regressed from 0.029 s/op to 0.050 s/op. I also noticed that if I refactor only 2 branches of code (all nulls and no filter), the benchmark doesn't show any regression.
> 
> Here is what I have now. Do you like this better? (I'm concerned that this shape would prompt readers to refactor the last branch and if they don't run the benchmark they won't know that it introduced a regression.)
> 
> ```
>         int streamPosition = 0;
>         if (dataStream == null && presentStream != null) {
>             streamPosition = readAllNulls(positions, positionCount);
>         }
>         else if (filter == null) {
>             streamPosition = readNoFilter(positions, positionCount);
>         }
>         else {
>             outputPositionCount = 0;
>             for (int i = 0; i < positionCount; i++) {
>                 int position = positions[i];
>                 if (position > streamPosition) {
>                     skip(position - streamPosition);
>                     streamPosition = position;
>                 }
> 
>                 if (presentStream != null && !presentStream.nextBit()) {
>                     if (nullsAllowed) {
>                         if (outputRequired) {
>                             nulls[outputPositionCount] = true;
>                         }
>                         outputPositions[outputPositionCount] = position;
>                         outputPositionCount++;
>                     }
>                 }
>                 else {
>                     boolean value = dataStream.nextBit();
>                     if (filter.testBoolean(value)) {
>                         if (outputRequired) {
>                             values[outputPositionCount] = (byte) (value ? 1 : 0);
>                             if (presentStream != null && nullsAllowed) {
>                                 nulls[outputPositionCount] = false;
>                             }
>                         }
>                         outputPositions[outputPositionCount] = position;
>                         outputPositionCount++;
>                     }
>                 }
>                 streamPosition++;
>             }
>         }
> ```

Yes that's fair.  I even looked at the decompiled bytecode to see if there's a difference at that level, and they're basically the same except for the `invokespecial` to the additional method which should be inlined.  Maybe we could add a comment explaining the reason for inlining the else branch, in the meantime I'll try to see if some profiling tools reveal the source of the inefficiency.

mbasmanova commented:
@tdcmeehan 

> Maybe we could add a comment explaining the reason for inlining the else branch

I added this comment:

```
            // This branch is inlined because extracting this code into a helper method (readWithFilter)
            // results in performance regressions on BenchmarkSelectiveStreamReaders.readBooleanNoNull[withFilter]
            // benchmarks.
            // The differences are:
            // - readBooleanNoNull: 0.045 vs.0.029 s/op
            // - readBooleanNoNullWithFilter: 0.112 vs 0.099 s/op
```

And here are full benchmark results:

```
# Run complete. Total time: 00:02:22

Benchmark                                                      (typeSignature)  Mode  Cnt  Score    Error  Units
BenchmarkSelectiveStreamReaders.readAllNull                            boolean  avgt   20  0.003 ±  0.001   s/op
BenchmarkSelectiveStreamReaders.readBooleanNoNull                          N/A  avgt   20  0.028 ±  0.001   s/op
BenchmarkSelectiveStreamReaders.readBooleanNoNullWithFilter                N/A  avgt   20  0.097 ±  0.002   s/op
BenchmarkSelectiveStreamReaders.readBooleanWithNull                        N/A  avgt   20  0.094 ±  0.001   s/op
BenchmarkSelectiveStreamReaders.readBooleanWithNullWithFilter              N/A  avgt   20  0.133 ±  0.001   s/op


# Run complete. Total time: 00:02:25

Benchmark                                                      (typeSignature)  Mode  Cnt  Score    Error  Units
BenchmarkSelectiveStreamReaders.readAllNull                            boolean  avgt   20  0.003 ±  0.001   s/op
BenchmarkSelectiveStreamReaders.readBooleanNoNull                          N/A  avgt   20  0.045 ±  0.001   s/op
BenchmarkSelectiveStreamReaders.readBooleanNoNullWithFilter                N/A  avgt   20  0.112 ±  0.003   s/op
BenchmarkSelectiveStreamReaders.readBooleanWithNull                        N/A  avgt   20  0.095 ±  0.002   s/op
BenchmarkSelectiveStreamReaders.readBooleanWithNullWithFilter              N/A  avgt   20  0.149 ±  0.003   s/op
```
